Transport encryption for ZMQ communication (#145)
jupyter/enhancement-proposals: 145-transport-encryption/transport-encryption.md

---
title: Transport encryption for ZMQ communication
authors: Benjamin Ragan-Kelley (@minrk), Michał Krassowski (@krassowski), Matthias Bussonnier (@Carreau)
issue-number: 75
pr-number: 145
date-started: 2026-06-10
---

# Transport encryption for ZMQ communication

## Summary

We propose to support encrypting the traffic exchanged between the kernels and client, adding a dedicated stanza in `kernelspec` to enable kernels to communicate support for encryption, adding a new set of fields for key exhange in the connection file, and using CurveZMQ protocol as a default implementation of such an encryption.

## Motivation

> With the default TCP transport non-authenticated clients with access to ports on the same machine are currently able to monitor IOPub, accessing any secrets exchanged during kernel runtime (code, outputs, etc). It can be only prevented by using ICP transport and file permissions.

Reviewer: ```suggestion
With the default TCP transport non-authenticated clients with access to ports on the same machine are currently able to monitor IOPub, accessing any secrets exchanged during kernel runtime (code, outputs, etc). It can be only prevented by using IPC transport and file permissions.
```

- Client, front-end, and server authors generally get this for free through `jupyter_client` ≥ 8.9 and `jupyter_server` ≥ 2.20; the manager generates the keys and writes them into the connection file.
- Operators flip one setting; the `auto` versus `required` distinction lets them roll encryption out gradually and then enforce it.

There are two new diagnostics a user may notice. When a kernel runs over TCP _without_ encryption, `ipykernel` now logs a warning:

> Kernel is running over TCP without encryption. All traffic, including any secrets in cell outputs, is readable by other processes on this machine. Use the IPC transport or start the kernel through a manager that provisions CurveZMQ keys.

And requesting `transport_encryption` of `required` for a kernel that does not advertise support fails at startup:

> RuntimeError: transport_encryption='required' but kernelspec does not declare `metadata.supported_encryption='curve'`.

> We recommend other kernels that decide to implement encryption to emit similar warning and error messages.

Reviewer: ```suggestion
We recommend other kernels and clients that decide to implement encryption to emit similar warning and error messages.
```

    M<<->>K: encrypted, authenticated ZMQ traffic
```

### Socket configuration

On the kernel (server) side, for every socket it binds (`shell`, `stdin`, and `control` as `ROUTER`, `iopub` as `XPUB`, and `heartbeat` as `ROUTER`), the kernel sets, _before_ binding:

```python
socket.curve_secretkey = secret   # from connection file
socket.curve_publickey = public   # from connection file
socket.curve_server = True
```

On the client side, for the connecting sockets the behavior differs slightly by channel:

- `shell`, `iopub`, `stdin`, and `control` reuse the keypair from the connection file as their own identity and set `curve_serverkey` to the kernel's public key (which authenticates the server):

  ```python
  socket.curve_secretkey = secret
  socket.curve_publickey = public
  socket.curve_serverkey = public
  ```

- The `heartbeat` client instead generates a fresh ephemeral keypair per socket and only needs the server's public key (`curve_serverkey`); it never receives the secret key. The heartbeat runs in its own ZMQ context and thread (so the GIL cannot stall it), and the curve options are therefore applied when its socket is created inside that thread.

> This "simplest version" (a single keypair used by both ends) choice from the variants considered in the pre-proposal is intentional. The connection file already carries the keypair, so reusing it avoids introducing a second key-distribution mechanism. `curve_serverkey` is what actually authenticates the connection; the rest configures encryption.

Reviewer: ```suggestion
This "simplest version" (a single keypair used by both ends) choice from the variants considered in the pre-proposal is intentional. The connection file already carries the keypair, so reusing it avoids introducing a second key-distribution mechanism. `curve_serverkey` authenticates the connection; the rest configures encryption.
```

### The `transport_encryption` policy

The `transport_encryption` setting is configurable on the kernel manager in both `jupyter_client` and `jupyter_server`, with the values `disabled` (the default), `auto`, and `required`. `jupyter_server` is a thin layer: when the policy is not `disabled`, it forwards the policy into the per-kernel launch; the actual key generation, kernelspec check, and connection-file writing happen in `jupyter_client`. Setting the policy to `auto` or `required` while the local `libzmq` lacks Curve support (`zmq.has("curve")` is false) is rejected at configuration time, so the misconfiguration surfaces immediately rather than at first kernel start.

> The kernel itself has no encryption toggle of its own; it is purely _data-driven_. It applies Curve options if and only if the connection file it loads contains _both_ `curve_publickey` and `curve_secretkey`. A file with only one of the two is treated as unencrypted.

Reviewer: > A file with only one of the two is treated as unencrypted.

Given that this is security-related, we should probably treat this invalid partial config situation as a fatal error, rather than running insecurely.

Author: ```suggestion
The kernel itself has no encryption toggle of its own; it is purely _data-driven_.
It should apply Curve options if the connection file it loads contains _both_ `curve_publickey` and `curve_secretkey`.
A file with only one of the two should lead to an error and the kernel refusing to connect.
```

Reviewer 2: Not a request for a change, but should we mention somewhere that kernels might at some point in the future refuse to start if there is not keys present in the file ?
I'm assuming we do want to have tcp+encytption be the default, and have no-encryption forbiden (unless you are debugging and actually go out of your ways to disable it)

Author: > Not a request for a change, but should we mention somewhere that kernels might at some point in the future refuse to start if there is not keys present in the file ?

I think this would be very sensible for some kernel wrappers designed to run in environments with sensitive data. The JEP already calls mandatory encryption out in "future possibilities", broadly saying we can consider granular enforcement:

> - _Mandatory encryption_ for remote or multi-tenant deployments, with per-channel or per-deployment policies once the remote story above is settled.

I will add another bullet point saying that kernels may in future refuse connection if these keys are not present.

Author: Done in 0451cdd abd dc4e8e7

### Kernelspec capability advertisement

> A kernel declares support with `metadata.supported_encryption`. The value is matched case-insensitively and may be either the string `"curve"` or a list containing `"curve"` (leaving room to advertise multiple schemes in future). The check happens _before_ the connection is established: capability cannot be discovered at runtime via `kernel_info_reply`, because by the time a reply could arrive the (unencrypted) connection already exists. This mirrors how JEP 66 gates the handshake pattern on the static `kernel_protocol_version` field.

Reviewer: Does it make sense to start with a list, rather than singleton string? Then we don't need to handle a change in type when a second item is added?

Author: Done in e1d495f

To implement this JEP we will reflect these additions in the official schemas (JEP 105/106). In the connection file (currently `connectionfile-v1.0.schema.json`, which composes a `kernel_network` and a `signature` group), the two Curve fields fit naturally as a new optional `encryption` group, following the precedent that the `signature_scheme` and `key` material lives in the connection file:

```json
{
  "curve_publickey": "b{E{TUJ[Zxl1)]<>W(c*bEff[.[K]BGJE&4iF=oM",
  "curve_secretkey": "<dQ^&)<Lp#8IUJrWoOpOZ5+xjs]UV$ShhdqSEW))"
}
```

Both keys are optional and MUST appear together; their absence means no transport encryption, which is the legacy behavior. This would bump the schema to a new versioned `$id` (for example `connectionfile-v1.1`).

### Backward compatibility and corner cases

- _Default off._ With `transport_encryption` set to `disabled`, no keys are generated, the kernel launch is identical to today, and the Curve fields are absent from the connection file. A kernel that loads such a file behaves exactly as before.
- `auto` with a non-advertising kernelspec silently starts the kernel unencrypted; `required` with such a kernelspec fails at startup.

> - _Transport._ Encryption applies to `tcp` only; `ipc://` connections rely on filesystem permissions as today, and `required` with a non-`tcp` transport is an error.

Reviewer: I'm not sure required + non-tcp should lead to an error.

1. I think IPC might be _more_ secure than curve in a number of situations, so we don't want secure config to be mutually exclusive, and
2. you can actually still use curve on other transports. It might be overkill, but it's fine

So I think it makes the most sense to either:

1. _ignore_ transport encryption on ipc, or
2. apply it the same as tcp

both of which make good sense and don't lead to errors.

Reviewer 2: I think I would encrypt on IPC just to avoid extra branches in the code and treat the transport as opaque. I'm also thinking that there might be permission issues or misconfiguration that make an IPC a potential way to peek at data.

Author: Done in 528c9e1. I will follow with PRs to update PoC implementations. Since this is additive, we do not need to wait with releases.

---
title: Transport encryption for ZMQ communication
authors: Benjamin Ragan-Kelley (@minrk), Michał Krassowski (@krassowski), Matthias Bussonnier (@Carreau)
issue-number: 75
pr-number: 145
date-started: 2026-06-10
---

# Transport encryption for ZMQ communication

## Summary

> We propose to support encrypting the traffic exchanged between the kernels and client, adding a dedicated stanza in `kernelspec` to enable kernels to communicate support for encryption, adding a new set of fields for key exhange in the connection file, and using CurveZMQ protocol as a default implementation of such an encryption.

Reviewer 2: ```suggestion
We propose to support encrypting the traffic exchanged between the kernels and client, adding a dedicated stanza in `kernelspec` to enable kernels to communicate support for encryption, adding a new set of fields for key exchange in the connection file, and using CurveZMQ protocol as a default implementation of such an encryption.
```

## Motivation

With the default TCP transport non-authenticated clients with access to ports on the same machine are currently able to monitor IOPub, accessing any secrets exchanged during kernel runtime (code, outputs, etc). It can be only prevented by using IPC transport and file permissions.

## Guide-level explanation

> The five ZMQ channels a kernel exposes (`shell`, `iopub`, `stdin`, `control`, and `heartbeat`) can now be wrapped in _transport encryption_: the connection between a client and a kernel is encrypted and mutually authenticated at the socket level, using ZeroMQ's [CurveZMQ](https://rfc.zeromq.org/spec/26/) protocol. CurveZMQ provides forward secrecy (it negotiates ephemeral per-connection session keys, so traffic captured today cannot be decrypted later even if the long-term keys leak) and connection-level authentication (a peer that does not hold the right key never gets its messages delivered, instead of being filtered out after the fact by an HMAC check).

Reviewer 2: what do you mean by mutual ? I'm not sure I understand if the client distribute the keys, it can't distinguish between many servers (kernels) using the same keys , woudl that be with ZAP.

(also personal preference I prefer hardwrap to not point to full paragraphs)

Author: Addressed in 6a1ecce.

Encryption is off by default and is controlled by a single operator-facing setting on the kernel manager, `transport_encryption`, which takes one of three values:

- `disabled` (the default): kernels start exactly as they do today, unencrypted; nothing changes.
- `auto`: a kernel is started encrypted _only if_ its kernelspec advertises that it understands encryption; kernels that do not are started unencrypted. This is the value an operator can safely set globally: encryption-capable kernels get encryption, everything else keeps working.
- `required`: every kernel is started encrypted, and starting a kernel whose kernelspec does _not_ advertise encryption support fails fast with a clear error rather than silently falling back to plaintext.

A kernel advertises that it understands encryption through a new key in its `kernel.json` metadata:

```json
{
  "argv": ["...", "{connection_file}"],
  "display_name": "Python 3 (ipykernel)",
  "language": "python",
  "metadata": {

>     "supported_encryption": ["curve"]

Reviewer 2: thinking at loud; not requests to change.

Should this be plural ?

Author: IMO `encryptions` sounds funny in English. See https://english.stackexchange.com/questions/260385/is-it-ever-acceptable-to-pluralize-the-word-encryption-to-encryptions. If we were pedantic, we would need to change it to `supported_encryption_schemas` or so which is a bit long. I think it is ok to keep as-is.

Reviewer 2: Fair. With a pure linguistic view it make sens.

How others should think about the feature:

- Kernel authors add one line to their `kernel.json` metadata (`"supported_encryption": ["curve"]`) and, if their kernel is not built on `ipykernel` or `jupyter_client`, set three socket options (see the reference section) on the sockets they bind. They do _not_ generate or manage keys.
- Client, front-end, and server authors generally get this for free through `jupyter_client` ≥ 8.9 and `jupyter_server` ≥ 2.20; the manager generates the keys and writes them into the connection file.
- Operators flip one setting; the `auto` versus `required` distinction lets them roll encryption out gradually and then enforce it.

There are two new diagnostics a user may notice. When a kernel runs over TCP _without_ encryption, `ipykernel` now logs a warning:

> Kernel is running over TCP without encryption. All traffic, including any secrets in cell outputs, is readable by other processes on this machine. Use the IPC transport or start the kernel through a manager that provisions CurveZMQ keys.

And requesting `transport_encryption` of `required` for a kernel that does not advertise support fails at startup:

> > RuntimeError: transport_encryption='required' but kernelspec does not declare `metadata.supported_encryption='curve'`.

Reviewer 2: Accurate, but I think the message should be updated to say 'curve' not in ..., 

also supported_encryption maybe should have been plural, but that ship might have sailed now. See above comment.

Author: ```suggestion
> RuntimeError: transport_encryption='required' but kernelspec does not declare 'curve' in `metadata.supported_encryption`.
```

- CurveZMQ ([ZMQ RFC 26](https://rfc.zeromq.org/spec/26/)) is an adaptation of Daniel J. Bernstein's CurveCP to ZeroMQ, layered over the ZMTP wire protocol; the related ZAP ([RFC 27](https://rfc.zeromq.org/spec/27/)) defines the authentication hook. These are mature, widely deployed specifications.
- JEP 66 (Kernel handshaking) establishes that the connection file is the right place to carry the signature scheme and key, and that capability gating happens via a static kernelspec field (`kernel_protocol_version` ≥ 5.5) decided by the launcher. This proposal follows both precedents.
- A known bad experience: a `libzmq` thread-safety bug around `curve_keypair` and `libsodium` on systems without `getrandom` ([zeromq/libzmq#4241](https://github.com/zeromq/libzmq/issues/4241)) was closed upstream as _wontfix_. Opt-in patches have been backported into `pyzmq`'s bundled `libzmq` and into conda-forge's `libzmq`, but this remains a residual risk to track.

## Unresolved questions

To be discussed during the JEP process:

### Key isolation

> The reference implementation reuses one keypair for both the kernel's server identity and the local clients' identity, so the _secret_ key sits in the connection file and is loaded by every local client. This matches the `session.key` trust model and gives forward secrecy against passive capture, but the long-term secret in the file is sufficient for a live man-in-the-middle. Should we instead distribute only the server _public_ key and have clients (and possibly the kernel) use distinct keypairs? The pre-proposal raised holding the secret in an environment variable (for example `$JUPYTER_CURVE_SECRETKEY`) so the connection file alone is no longer enough to impersonate either end. As of now this JEP leaves standardizing it to a follow-up work, but should it be deemded universally desirable, we could support alternative key distribution mechanisms.

Reviewer 2: ```suggestion
The reference implementation reuses one keypair for both the kernel's server identity and the local clients' identity, so the _secret_ key sits in the connection file and is loaded by every local client. This matches the `session.key` trust model and gives forward secrecy against passive capture, but the long-term secret in the file is sufficient for a live man-in-the-middle. Should we instead distribute only the server _public_ key and have clients (and possibly the kernel) use distinct keypairs? The pre-proposal raised holding the secret in an environment variable (for example `$JUPYTER_CURVE_SECRETKEY`) so the connection file alone is no longer enough to impersonate either end. As of now this JEP leaves standardizing it to a follow-up work, but should it be deemed universally desirable, we could support alternative key distribution mechanisms.
```

### HMAC message signing

Connection-level authentication can in principle remove the need for per-message HMAC verification, but the current implementation keeps signing as-is. For now we propose to keep it as defense-in-depth, but one could argue that opt-out of HMAC should be a part of this proposal to fully realise the promised benefits. While this is not currently part of the submitted PoCs, this JEP should encourage power users to experiment with opt-out, which may reduce the overhead of execution.

### Connection-file schema

> As of now the added fileds are `curve_publickey` and `curve_secretkey` but we could consider a dedicated `encryption` group before we codify that in the schema.

Reviewer 2: ```suggestion
As of now the added fields are `curve_publickey` and `curve_secretkey` but we could consider a dedicated `encryption` group before we codify that in the schema.
```

### Out of scope

Considered out of scope for this JEP (addressable independently later):

- _Remote and gateway kernels._ The server toggle is not propagated to gateway-managed kernels, and remote provisioners that ship the connection file over the network introduce a key-bootstrapping problem of their own. How encryption composes with remote kernels and with the JEP 66 registration-socket handshake is left to follow-up work.
- _Key rotation._ Keys are fixed for the lifetime of a connection file (including across restarts); a rotation mechanism is out of scope here.

## Future possibilities

> - _Handshake key exchange._ Rather than writing the secret into a shared connection file, encryption keys could be exchanged over the registration socket during the JEP 66 handshake, which would also make remote key exchange tractable without persisting secrets to disk.

Reviewer 3: Should we handle the handshake pattern in this JEP? The keys should be communicated via the initial connection file (the one passing the registration socket info), since the kernel would bind its socket before communicating them back to the servef / client.

Author: _How_ would like to handle it in the JEP? Add another `curve-handshake` option and let kernels decide what they support? Change the default to use handshake pattern?

Reviewer 3: I think it should be very similar to the usual connection pattern; the kernel advertise what it supports in its kernelspec, then if the client decides to initiate a handshake, and if the kernel supports Curve encryption, the client passes the keys in the registration file.

Author: I think there are three approaches:
- a) the "initial connection info" that is used to communicate where the registration socket is could contain the curve keys (i.e. for local provisioner they get persisted on disk) - no change needed in this JEP maybe beyond a single sentence clarification - it is trivial and gives us no advantage
- b) the curve keys are sent over the registration socket - this way we do not need to persist the curve on disk; the downside is we would then also have a separate set of short-lived keys for registration socket so that one cannot eavesdrop on that connection either
- c) we move the key generation to kernels; the kernel generates the curve key pair and sends _only_ the public key thorough the registration socket; this is possibly the best future state (and my initial PoC included kernels generating the keys) but has a steeper cost for adoption: kernels need to generate their own keys (so be built with libsodium support for zmq), we need to have JEP 66 implemented etc.

The line 244 is saying "in future we may want to consider (b) or (c)". Are you suggesting that for now we just document (a)?

Reviewer: I think it's fine to leave handshake/registration as future work. If folks think it needs to be addressed here, we should use a curve key _on the registration socket itself_, since registration in the current handshake suffers from the same issues motivating curve in the first place.

Reviewer 3: > The line 244 is saying "in future we may want to consider (b) or (c)". Are you suggesting that for now we just document (a)?

Yes, I think (a) can be part of this JEP,  and (b) and (c) left as possible future work.

Author: I pushed 4ad2d85883d4d9cc09ea20d02294f6b666d90275 let me know if I missed something in there. Otherwise, I think this is ready for a vote.

Reviewer 3: I think this is ready for a vote.
